File: imageStreamer.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/compressed_image.hpp"

#include <websocketpp/config/asio_no_tls.hpp>
#include <websocketpp/server.hpp>
#include <nlohmann/json.hpp>

#include <atomic>
#include <cstring>
#include <map>
#include <memory>
#include <mutex>
#include <set>
#include <thread>
#include <vector>

typedef websocketpp::server<websocketpp::config::asio> Server;
using websocketpp::connection_hdl;
using json = nlohmann::json;

#define GREEN  "\033[1;32m"
#define RED    "\033[1;31m"
#define YELLOW "\033[1;33m"
#define CYAN   "\033[1;36m"
#define CLEAR  "\033[0m"

// Macros de logging — só funcionam dentro da classe (usam this->get_logger())
#define INFO(msg)  RCLCPP_INFO_STREAM (this->get_logger(), CYAN   << msg << CLEAR)
#define OK(msg)    RCLCPP_INFO_STREAM (this->get_logger(), GREEN  << msg << CLEAR)
#define WARN(msg)  RCLCPP_WARN_STREAM (this->get_logger(), YELLOW << msg << CLEAR)
#define ERROR(msg) RCLCPP_ERROR_STREAM(this->get_logger(), RED    << msg << CLEAR)

// ── Data Structures ──────────────────────────────────────────────────────────

struct ClientState {
  int         id;
  std::string activeTopic;
};

struct TopicState {
  // ROS2: SharedPtr em vez de ros::Subscriber
  rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr subscription;
  std::set<connection_hdl, std::owner_less<connection_hdl>> clients;
};

// ── ImageStreamer Node ────────────────────────────────────────────────────────

class ImageStreamer : public rclcpp::Node   // ← herda de rclcpp::Node
{
public:
  ImageStreamer(int argc, char **argv);
  ~ImageStreamer();

private:
  std::atomic<int> nextClientID{1};
  std::map<connection_hdl, ClientState, std::owner_less<connection_hdl>> clientStates;
  std::map<std::string, TopicState> topicStates;
  std::mutex stateMutex;

  Server            wsServer;
  std::thread       wsThread;
  std::atomic<bool> wsRunning{false};
  int               socketPort{9092};

  int  getClientID(connection_hdl hdl);
  void messageHandler  (connection_hdl hdl, Server::message_ptr msg);
  void handleStreamEnable (connection_hdl hdl, const std::string &topic);
  void handleStreamDisable(connection_hdl hdl, const std::string &topic);
  void handleStreamSwitch (connection_hdl hdl, const std::string &newTopic);
  void imageCallback(sensor_msgs::msg::CompressedImage::ConstSharedPtr msg,
                     const std::string &topic);
};

// ── Constructor ───────────────────────────────────────────────────────────────

ImageStreamer::ImageStreamer(int argc, char **argv)
  : rclcpp::Node("gui_image_streamer")   // ← inicializa Node
{
  // Argumento CLI para a porta
  if (argc >= 2) {
    try   { socketPort = std::stoi(argv[1]); }
    catch (const std::exception &e) {
      WARN("[ImageStreamer] Porta CLI inválida '" << argv[1] << "': " << e.what());
    }
  }

  // Parâmetro ROS2 (sobrepõe CLI se definido)
  this->declare_parameter("socket_port", socketPort);
  socketPort = this->get_parameter("socket_port").as_int();

  // WebSocket Server
  INFO("[ImageStreamer] A iniciar na porta " << socketPort);
  wsServer.init_asio();
  wsServer.set_reuse_addr(true);
  wsServer.clear_access_channels(websocketpp::log::alevel::all);
  wsServer.clear_error_channels(websocketpp::log::elevel::all);

  wsServer.set_open_handler([this](connection_hdl hdl) {
    int id = nextClientID++;
    std::size_t total;
    { std::lock_guard<std::mutex> lock(stateMutex);
      clientStates[hdl] = ClientState{id, ""}; total = clientStates.size(); }
    INFO("[ImageStreamer] Cliente " << id << " ligado. Total: " << total);
  });

  wsServer.set_close_handler([this](connection_hdl hdl) {
    std::string activeTopic;
    int clientId = -1;
    std::size_t total = 0, remaining = 0;
    rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr subToReset;

    { std::lock_guard<std::mutex> lock(stateMutex);
      auto cIt = clientStates.find(hdl);
      if (cIt != clientStates.end()) {
        clientId = cIt->second.id; activeTopic = cIt->second.activeTopic;
        clientStates.erase(cIt);
      }
      total = clientStates.size();
      if (!activeTopic.empty()) {
        auto tIt = topicStates.find(activeTopic);
        if (tIt != topicStates.end()) {
          tIt->second.clients.erase(hdl);
          remaining = tIt->second.clients.size();
          if (tIt->second.clients.empty()) {
            subToReset = tIt->second.subscription;
            topicStates.erase(tIt);
          }
        }
      }
    }

    subToReset.reset(); // ← ROS2: reset do SharedPtr = unsubscribe

    if (!activeTopic.empty())
      INFO("[ImageStreamer] Cliente " << clientId << " desligado de '" << activeTopic
           << "'. Restantes: " << remaining << ". Total: " << total);
    else
      INFO("[ImageStreamer] Cliente " << clientId << " desligado. Total: " << total);
  });

  wsServer.set_message_handler([this](connection_hdl hdl, Server::message_ptr msg) {
    this->messageHandler(hdl, msg);
  });

  wsThread = std::thread([this]() {
    try {
      wsServer.listen(socketPort);
      wsServer.start_accept();
      wsRunning = true;
      wsServer.run();
    } catch (const std::exception &e) {
      ERROR("[ImageStreamer] Erro no servidor: " << e.what());
      rclcpp::shutdown(); // ← ROS2
    }
    wsRunning = false;
  });
}

// ── Destructor ────────────────────────────────────────────────────────────────

ImageStreamer::~ImageStreamer()
{
  std::vector<connection_hdl> openClients;
  std::vector<rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr> subsToClear;

  { std::lock_guard<std::mutex> lock(stateMutex);
    for (const auto &e : clientStates)  openClients.push_back(e.first);
    for (auto &e : topicStates)         subsToClear.push_back(e.second.subscription);
    clientStates.clear();
    topicStates.clear();
  }

  subsToClear.clear(); // destrói subscriptions

  for (const auto &hdl : openClients) {
    try { wsServer.close(hdl, websocketpp::close::status::going_away, "shutdown"); }
    catch (...) {}
  }

  try { wsServer.stop_listening(); } catch (...) {}
  if (wsRunning) wsServer.stop();
  if (wsThread.joinable()) wsThread.join();
}

// ── Utility ───────────────────────────────────────────────────────────────────

int ImageStreamer::getClientID(connection_hdl hdl)
{
  std::lock_guard<std::mutex> lock(stateMutex);
  auto it = clientStates.find(hdl);
  return (it == clientStates.end()) ? -1 : it->second.id;
}

// ── Message Handler ───────────────────────────────────────────────────────────

void ImageStreamer::messageHandler(connection_hdl hdl, Server::message_ptr msg)
{
  if (msg->get_opcode() != websocketpp::frame::opcode::text) return;

  const std::string payload = msg->get_payload();
  if (payload.size() < 2) return;

  json j;
  try { j = json::parse(payload); }
  catch (...) { ERROR("[ImageStreamer] Erro a parsear JSON"); return; }

  if (!j.contains("cmd") || !j.contains("topic") ||
      !j["cmd"].is_string() || !j["topic"].is_string()) return;

  const std::string cmd = j["cmd"], topic = j["topic"];
  if (topic.empty() || topic[0] != '/') return;

  if      (cmd == "enable")  handleStreamEnable(hdl, topic);
  else if (cmd == "disable") handleStreamDisable(hdl, topic);
  else if (cmd == "switch")  handleStreamSwitch(hdl, topic);
}

// ── Stream Enable ─────────────────────────────────────────────────────────────

void ImageStreamer::handleStreamEnable(connection_hdl hdl, const std::string &topic)
{
  int clientId; std::size_t viewers;
  { std::lock_guard<std::mutex> lock(stateMutex);
    auto cIt = clientStates.find(hdl);
    if (cIt == clientStates.end()) return;
    clientId = cIt->second.id;
    if (!cIt->second.activeTopic.empty()) {
      WARN("[ImageStreamer] Cliente " << clientId << " já está a ver um tópico. Usa 'switch'.");
      return;
    }
    auto tIt = topicStates.find(topic);
    if (tIt == topicStates.end()) {
      TopicState ts;
      // ROS2: create_subscription com SensorDataQoS (best_effort, depth 1)
      ts.subscription = this->create_subscription<sensor_msgs::msg::CompressedImage>(
          topic, rclcpp::SensorDataQoS(),
          [this, topic](sensor_msgs::msg::CompressedImage::ConstSharedPtr m) {
            this->imageCallback(m, topic);
          });
      tIt = topicStates.emplace(topic, std::move(ts)).first;
    }
    tIt->second.clients.insert(hdl);
    cIt->second.activeTopic = topic;
    viewers = tIt->second.clients.size();
  }
  OK("[ImageStreamer] Cliente " << clientId << " -> '" << topic << "'. Viewers: " << viewers);
}

// ── Stream Disable ────────────────────────────────────────────────────────────

void ImageStreamer::handleStreamDisable(connection_hdl hdl, const std::string &topic)
{
  bool shouldReset = false;
  rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr subToReset;
  int clientId; std::size_t viewers = 0;

  { std::lock_guard<std::mutex> lock(stateMutex);
    auto cIt = clientStates.find(hdl);
    if (cIt == clientStates.end() || cIt->second.activeTopic != topic) return;
    clientId = cIt->second.id;
    auto tIt = topicStates.find(topic);
    if (tIt != topicStates.end()) {
      tIt->second.clients.erase(hdl);
      viewers = tIt->second.clients.size();
      if (tIt->second.clients.empty()) {
        subToReset = tIt->second.subscription; shouldReset = true;
        topicStates.erase(tIt);
      }
    }
    cIt->second.activeTopic.clear();
  }

  if (shouldReset) subToReset.reset();
  OK("[ImageStreamer] Cliente " << clientId << " -> Disabled '" << topic << "'. Viewers: " << viewers);
}

// ── Stream Switch ─────────────────────────────────────────────────────────────

void ImageStreamer::handleStreamSwitch(connection_hdl hdl, const std::string &newTopic)
{
  std::string oldTopic;
  rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr oldSubToReset;
  bool resetOld = false;
  int clientId; std::size_t newViewers = 0, oldViewers = 0;

  { std::lock_guard<std::mutex> lock(stateMutex);
    auto cIt = clientStates.find(hdl);
    if (cIt == clientStates.end()) return;
    clientId = cIt->second.id; oldTopic = cIt->second.activeTopic;
    if (oldTopic == newTopic) return;

    auto nIt = topicStates.find(newTopic);
    if (nIt == topicStates.end()) {
      TopicState ts;
      ts.subscription = this->create_subscription<sensor_msgs::msg::CompressedImage>(
          newTopic, rclcpp::SensorDataQoS(),
          [this, newTopic](sensor_msgs::msg::CompressedImage::ConstSharedPtr m) {
            this->imageCallback(m, newTopic);
          });
      nIt = topicStates.emplace(newTopic, std::move(ts)).first;
    }
    nIt->second.clients.insert(hdl); newViewers = nIt->second.clients.size();
    cIt->second.activeTopic = newTopic;

    if (!oldTopic.empty()) {
      auto oIt = topicStates.find(oldTopic);
      if (oIt != topicStates.end()) {
        oIt->second.clients.erase(hdl); oldViewers = oIt->second.clients.size();
        if (oIt->second.clients.empty()) {
          oldSubToReset = oIt->second.subscription; resetOld = true;
          topicStates.erase(oIt);
        }
      }
    }
  }

  if (resetOld) oldSubToReset.reset();
  OK("[ImageStreamer] Cliente " << clientId << " -> '"
     << (oldTopic.empty() ? "none" : oldTopic) << "' → '" << newTopic
     << "'. Viewers novo: " << newViewers << ", antigo: " << oldViewers);
}

// ── Image Callback ────────────────────────────────────────────────────────────

void ImageStreamer::imageCallback(
    sensor_msgs::msg::CompressedImage::ConstSharedPtr msg,
    const std::string &topic)
{
  std::vector<connection_hdl> interested;
  { std::lock_guard<std::mutex> lock(stateMutex);
    auto tIt = topicStates.find(topic);
    if (tIt == topicStates.end() || tIt->second.clients.empty()) return;
    for (const auto &hdl : tIt->second.clients) interested.push_back(hdl);
  }

  json header;
  header["topic"]     = topic;
  header["format"]    = msg->format;
  header["timestamp"] = msg->header.stamp.sec + msg->header.stamp.nanosec * 1e-9; // ← ROS2

  const std::string headerStr  = header.dump();
  const uint32_t    headerSize = static_cast<uint32_t>(headerStr.size());
  const std::size_t totalSize  = sizeof(headerSize) + headerStr.size() + msg->data.size();

  std::vector<uint8_t> frame(totalSize);
  std::memcpy(frame.data(),                                       &headerSize,       sizeof(headerSize));
  std::memcpy(frame.data() + sizeof(headerSize),                  headerStr.data(),  headerStr.size());
  std::memcpy(frame.data() + sizeof(headerSize) + headerStr.size(), msg->data.data(), msg->data.size());

  for (const auto &hdl : interested) {
    try {
      wsServer.send(hdl, frame.data(), frame.size(), websocketpp::frame::opcode::binary);
    } catch (const std::exception &e) {
      ERROR("[ImageStreamer] Falha ao enviar para cliente " << getClientID(hdl) << ": " << e.what());
    }
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  // ROS2: MultiThreadedExecutor (equivalente ao AsyncSpinner(2) do ROS1)
  rclcpp::executors::MultiThreadedExecutor executor(rclcpp::ExecutorOptions(), 2);
  auto node = std::make_shared<ImageStreamer>(argc, argv);
  executor.add_node(node);
  executor.spin();

  rclcpp::shutdown();
  return 0;
}

File: image_streamer.py
#!/usr/bin/env python3
"""
image_streamer.py — ROS2 WebSocket Image Streamer (Python)
Tradução de imageStreamer.cpp (C++ + websocketpp) para Python (rclpy + websockets).

Protocolo binário idêntico ao original (compatível com imageFeedService.js):
  [4B uint32 LE = headerSize][JSON header: {topic, format, timestamp}][bytes da imagem]

Dependência extra (além do ROS2):
  pip3 install websockets
"""

import asyncio
import json
import struct
import sys
import threading

import rclpy
import rclpy.executors
from rclpy.node import Node
from rclpy.qos import (
    QoSProfile,
    QoSReliabilityPolicy,
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
)
from sensor_msgs.msg import CompressedImage

import websockets
import websockets.exceptions

# ── Cores ANSI para o terminal ─────────────────────────────────────────────────
GREEN  = "\033[1;32m"
RED    = "\033[1;31m"
YELLOW = "\033[1;33m"
CYAN   = "\033[1;36m"
CLEAR  = "\033[0m"


# ── Estruturas de dados ────────────────────────────────────────────────────────
# Equivalentes a ClientState e TopicState do C++.
# Guardados em dicts simples; protegidos por _lock.
#
# _client_states : { ws → {"id": int, "active_topic": str} }
# _topic_states  : { topic_str → {"subscription": rclpy.Subscription,
#                                  "clients": set[ws]} }


class ImageStreamer(Node):
    """
    Nó ROS2 que subscreve tópicos CompressedImage e os transmite via WebSocket.
    O servidor WebSocket corre num thread separado com asyncio.
    """

    def __init__(self):
        super().__init__("gui_image_streamer")

        # ── Parâmetro ROS2: porta do servidor ──────────────────────────────────
        self.declare_parameter("socket_port", 9092)
        self._port: int = (
            self.get_parameter("socket_port").get_parameter_value().integer_value
        )

        # Argumento CLI sobrepõe o parâmetro ROS2 (comportamento igual ao C++)
        if len(sys.argv) >= 2:
            try:
                self._port = int(sys.argv[1])
            except ValueError:
                self.get_logger().warn(
                    f"{YELLOW}[ImageStreamer] Porta CLI inválida '{sys.argv[1]}'"
                    f", a usar {self._port}{CLEAR}"
                )

        # ── Estado partilhado (threads ROS + asyncio) ──────────────────────────
        self._lock = threading.Lock()
        self._next_id: int = 1
        self._client_states: dict = {}   # ws → {"id", "active_topic"}
        self._topic_states: dict  = {}   # topic → {"subscription", "clients"}

        # ── Event loop asyncio num thread dedicado ────────────────────────────
        self._loop = asyncio.new_event_loop()
        self._ws_thread = threading.Thread(
            target=self._run_ws_server, daemon=True, name="ws_thread"
        )
        self._ws_thread.start()

        self.get_logger().info(
            f"{CYAN}[ImageStreamer] A iniciar na porta {self._port}{CLEAR}"
        )

    # ── WebSocket server ───────────────────────────────────────────────────────

    def _run_ws_server(self):
        """Corre o event loop asyncio com o servidor WebSocket (thread separado)."""
        asyncio.set_event_loop(self._loop)
        self._loop.run_until_complete(self._serve())

    async def _serve(self):
        """Abre o servidor e fica à escuta para sempre."""
        async with websockets.serve(self._handle_client, "0.0.0.0", self._port):
            self.get_logger().info(
                f"{GREEN}[ImageStreamer] WebSocket a escutar na porta {self._port}{CLEAR}"
            )
            await asyncio.Future()  # bloqueia até o nó ser destruído

    async def _handle_client(self, ws):
        """Callback chamado para cada nova ligação WebSocket."""
        # Regista o cliente (equivalente a wsServer.set_open_handler no C++)
        with self._lock:
            client_id = self._next_id
            self._next_id += 1
            self._client_states[ws] = {"id": client_id, "active_topic": ""}
            total = len(self._client_states)

        self.get_logger().info(
            f"{CYAN}[ImageStreamer] Cliente {client_id} ligado. Total: {total}{CLEAR}"
        )

        try:
            # Loop de receção de mensagens
            async for message in ws:
                if isinstance(message, str):
                    await self._handle_message(ws, message)
                # Mensagens binárias ignoradas — o cliente não envia binário
        except websockets.exceptions.ConnectionClosed:
            pass
        finally:
            # Limpeza ao desligar (equivalente a wsServer.set_close_handler)
            await self._cleanup_client(ws)

    async def _handle_message(self, ws, payload: str):
        """
        Processa um comando JSON enviado pelo cliente.
        Protocolo (texto):
          { "cmd": "enable",  "topic": "/camera/compressed" }
          { "cmd": "disable", "topic": "/camera/compressed" }
          { "cmd": "switch",  "topic": "/camera2/compressed" }
        """
        if len(payload) < 2:
            return

        try:
            j = json.loads(payload)
        except json.JSONDecodeError:
            self.get_logger().error(
                f"{RED}[ImageStreamer] Erro a parsear JSON: {payload[:80]}{CLEAR}"
            )
            return

        cmd   = j.get("cmd",   "")
        topic = j.get("topic", "")

        if not isinstance(cmd, str) or not isinstance(topic, str):
            return
        if not topic or not topic.startswith("/"):
            return

        # Despacha para o handler correto (mesmo que messageHandler() no C++)
        if   cmd == "enable":  self._handle_enable(ws, topic)
        elif cmd == "disable": self._handle_disable(ws, topic)
        elif cmd == "switch":  self._handle_switch(ws, topic)

    async def _cleanup_client(self, ws):
        """
        Remove o cliente dos estados e cancela a subscrição ROS2 se ficar vazia.
        Equivalente a wsServer.set_close_handler no C++.
        """
        sub_to_destroy = None
        client_id      = -1
        active_topic   = ""
        remaining      = 0
        total          = 0

        with self._lock:
            state = self._client_states.pop(ws, None)
            if state:
                client_id    = state["id"]
                active_topic = state["active_topic"]
            total = len(self._client_states)

            if active_topic:
                ts = self._topic_states.get(active_topic)
                if ts:
                    ts["clients"].discard(ws)
                    remaining = len(ts["clients"])
                    if not ts["clients"]:
                        sub_to_destroy = ts["subscription"]
                        del self._topic_states[active_topic]

        # destroy_subscription deve ser chamado fora do lock
        if sub_to_destroy:
            self._destroy_sub(sub_to_destroy)

        if active_topic:
            self.get_logger().info(
                f"{CYAN}[ImageStreamer] Cliente {client_id} desligado de "
                f"'{active_topic}'. Restantes: {remaining}. Total: {total}{CLEAR}"
            )
        else:
            self.get_logger().info(
                f"{CYAN}[ImageStreamer] Cliente {client_id} desligado. Total: {total}{CLEAR}"
            )

    # ── Comandos de stream ────────────────────────────────────────────────────
    # Equivalentes a handleStreamEnable/Disable/Switch no C++.
    # Chamados a partir do thread asyncio mas protegidos pelo _lock.

    def _handle_enable(self, ws, topic: str):
        """Começa a enviar o tópico para este cliente."""
        sub_created = None
        client_id   = -1
        viewers     = 0

        with self._lock:
            state = self._client_states.get(ws)
            if not state:
                return
            client_id = state["id"]

            if state["active_topic"]:
                self.get_logger().warn(
                    f"{YELLOW}[ImageStreamer] Cliente {client_id} já está a ver "
                    f"'{state['active_topic']}'. Usa 'switch'.{CLEAR}"
                )
                return

            # Cria subscrição ROS2 se ainda não existir para este tópico
            if topic not in self._topic_states:
                sub = self._create_sub(topic)
                self._topic_states[topic] = {"subscription": sub, "clients": set()}

            self._topic_states[topic]["clients"].add(ws)
            state["active_topic"] = topic
            viewers = len(self._topic_states[topic]["clients"])

        self.get_logger().info(
            f"{GREEN}[ImageStreamer] Cliente {client_id} -> '{topic}'. "
            f"Viewers: {viewers}{CLEAR}"
        )

    def _handle_disable(self, ws, topic: str):
        """Para de enviar o tópico para este cliente."""
        sub_to_destroy = None
        client_id      = -1
        viewers        = 0

        with self._lock:
            state = self._client_states.get(ws)
            if not state or state["active_topic"] != topic:
                return
            client_id = state["id"]

            ts = self._topic_states.get(topic)
            if ts:
                ts["clients"].discard(ws)
                viewers = len(ts["clients"])
                if not ts["clients"]:
                    sub_to_destroy = ts["subscription"]
                    del self._topic_states[topic]

            state["active_topic"] = ""

        if sub_to_destroy:
            self._destroy_sub(sub_to_destroy)

        self.get_logger().info(
            f"{GREEN}[ImageStreamer] Cliente {client_id} -> Disabled "
            f"'{topic}'. Viewers: {viewers}{CLEAR}"
        )

    def _handle_switch(self, ws, new_topic: str):
        """Troca atomicamente de tópico (sem reconexão WebSocket)."""
        old_sub_to_destroy = None
        client_id          = -1
        old_topic          = ""
        new_viewers        = 0
        old_viewers        = 0

        with self._lock:
            state = self._client_states.get(ws)
            if not state:
                return
            client_id = state["id"]
            old_topic = state["active_topic"]

            if old_topic == new_topic:
                return  # já está no tópico pedido

            # Subscreve o novo tópico se ainda não existir
            if new_topic not in self._topic_states:
                sub = self._create_sub(new_topic)
                self._topic_states[new_topic] = {"subscription": sub, "clients": set()}

            self._topic_states[new_topic]["clients"].add(ws)
            new_viewers = len(self._topic_states[new_topic]["clients"])
            state["active_topic"] = new_topic

            # Remove do tópico antigo
            if old_topic:
                ts = self._topic_states.get(old_topic)
                if ts:
                    ts["clients"].discard(ws)
                    old_viewers = len(ts["clients"])
                    if not ts["clients"]:
                        old_sub_to_destroy = ts["subscription"]
                        del self._topic_states[old_topic]

        if old_sub_to_destroy:
            self._destroy_sub(old_sub_to_destroy)

        self.get_logger().info(
            f"{GREEN}[ImageStreamer] Cliente {client_id} -> "
            f"'{'none' if not old_topic else old_topic}' → '{new_topic}'. "
            f"Viewers novo: {new_viewers}, antigo: {old_viewers}{CLEAR}"
        )

    # ── Helpers de subscrição ROS2 ────────────────────────────────────────────

    def _create_sub(self, topic: str):
        """
        Cria uma subscrição ROS2 para CompressedImage.
        QoS equivalente a SensorDataQoS do C++: best_effort, volatile, depth=1.
        """
        qos = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1,
        )
        return self.create_subscription(
            CompressedImage,
            topic,
            lambda msg, t=topic: self._image_callback(msg, t),
            qos,
        )

    def _destroy_sub(self, sub) -> None:
        """Destrói uma subscrição ROS2 de forma segura."""
        try:
            self.destroy_subscription(sub)
        except Exception as e:
            self.get_logger().warn(
                f"{YELLOW}[ImageStreamer] Erro ao destruir subscrição: {e}{CLEAR}"
            )

    # ── Image callback (thread do executor ROS2) ──────────────────────────────

    def _image_callback(self, msg: CompressedImage, topic: str) -> None:
        """
        Chamado pelo executor ROS2 a cada nova frame.
        Constrói o frame binário e envia para todos os clientes interessados.

        Frame layout (idêntico ao C++):
          [4B uint32 LE headerSize][JSON header bytes][bytes da imagem]
        """
        # Recolhe clientes interessados neste tópico
        with self._lock:
            ts = self._topic_states.get(topic)
            if not ts or not ts["clients"]:
                return
            interested = list(ts["clients"])

        # Constrói o header JSON (equivalente ao nlohmann::json no C++)
        header = {
            "topic":     topic,
            "format":    msg.format,
            "timestamp": msg.header.stamp.sec + msg.header.stamp.nanosec * 1e-9,
        }
        header_bytes = json.dumps(header, separators=(",", ":")).encode("utf-8")
        # uint32 little-endian (equivalente a memcpy de headerSize no C++)
        prefix = struct.pack("<I", len(header_bytes))
        frame  = prefix + header_bytes + bytes(msg.data)

        # Agenda o envio no loop asyncio (thread-safe: run_coroutine_threadsafe)
        for ws in interested:
            asyncio.run_coroutine_threadsafe(
                self._send_frame(ws, frame), self._loop
            )

    async def _send_frame(self, ws, frame: bytes) -> None:
        """Envia um frame binário para um cliente. Corre no loop asyncio."""
        try:
            await ws.send(frame)
        except (websockets.exceptions.ConnectionClosed, Exception) as e:
            # Cliente desligou entretanto — ignorado silenciosamente
            with self._lock:
                state = self._client_states.get(ws)
                cid   = state["id"] if state else "?"
            self.get_logger().error(
                f"{RED}[ImageStreamer] Falha ao enviar para cliente {cid}: {e}{CLEAR}"
            )


# ── Main ───────────────────────────────────────────────────────────────────────

def main():
    print("[ImageStreamer] A iniciar...", flush=True)
    rclpy.init()
    node = ImageStreamer()

    # MultiThreadedExecutor com 2 threads (equivalente ao C++)
    executor = rclpy.executors.MultiThreadedExecutor(num_threads=2)
    executor.add_node(node)

    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()